cmd/proxy/main.go

```diff
@@ -0,0 +1,64 @@
+package main
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"net"
+	"strings"
+
+	"github.com/spf13/viper"
+	"google.golang.org/grpc"
+
+	pb "github.com/improbable-eng/etcd-cluster-operator/api/proxy"
+)
+
+type proxyServer struct {
+	pb.UnimplementedProxyServer
+}
+
+func (ps *proxyServer) Backup(ctx context.Context, request *pb.BackupRequest) (*pb.BackupReply, error) {
+	return nil, nil
+}
+
+func (ps *proxyServer) Restore(ctx context.Context, request *pb.RestoreRequest) (*pb.RestoreReply, error) {
+	return nil, nil
+}
+
+func main() {
+	log.Printf("etcd-cluster-controller upload/download Proxy service")
+
+	// Setup defaults for expected configuration keys
+	viper.SetDefault("api-port", 8080)
+
+	// Allow configuration to be passed via environment variables
+	viper.SetEnvKeyReplacer(strings.NewReplacer("_", "-"))
+	viper.AutomaticEnv()
+
+	// Attempt to load a config file for configuration too
+	viper.SetConfigName("config")
+	viper.SetConfigType("yaml")
+	viper.AddConfigPath("/etc/etcd-backup-restore-proxy/")
```

Comment:
Do we need to use viper for all this? It seems a bit heavy for a very simple service, and it isn't an existing dependency of the project.

For now, can we stick to just using a struct with an `AddFlags(fs *flag.FlagSet)` method? (potentially using `pflags` if you need the additional functionality it provides)

Reply:
I also don't think we need to accept a config file at this point either - flags alone should be sufficient IMO.

Reply:
> it isn't an existing dependency of the project.

It kind of is, in the sense that the restore branch uses it. That's parked for now but will be coming back (likely with Viper) once I've updated it to use this proxy service.

> It seems a bit heavy for a very simple service,

Maybe this is personal preference, but I'd see *not* using Viper as heavier than using it. In the sense that it's more complex to wrangle the low-level environment variable options yourself, especially when it comes to defaulting.

> I also don't think we need to accept a config file at this point either - flags alone should be sufficient IMO.

I'm unsure on this. My expectation wasn't to use flags at all (the code in this PR doesn't). But use environment or a config file. If we don't use a config file and did use flags like you suggest then that's both evnvars and flags, which is a compelling reason to use Viper.

Additionally, Right now we only have the api-port option, but ultimately we're going to have a number of configuration options. At least:

* API port
* Metrics Port
* Metrics Path
* Google credentials file location
* Amazon credentials file location
* Maximum transfer size configuration for gRPC
* etc.

Reply:
I get that we might expand it out, but viper is _not_ just a magic 'it works' solution, you end out having to call into viper quite a bit to configure config loading, flag/env handling etc.

To your points, I:

a) don't think we need to support env based configuration
b) defaulting is totally handled as part of `flag.String`, `flag.Bool` etc
c) if we are going to allow config based configuration, we then lead onto how we document the schema for that file, which becomes a whole other thing to do. Kubernetes actually has component-base which aims to 'unify' how config files for Kubernetes components are provided (i.e. standardised names and some standard structures). I'd like to think if we _do_ need to expand out to support a config file, we'd look at using that as it'll make our operator configurable in a similar way to others 😄 

Overall I'm not _that_ fussed, but I don't agree that it is more complex to just use flags for now, and I don't personally see a need for a configuration file/structured schema or environment based configuration. All the examples you've provided above are either `string` or `int` 🤷‍♂

Reply:
(for reference, the cert-manager-webhook component's config: https://github.com/jetstack/cert-manager/blob/master/cmd/webhook/main.go#L56-L62)

Reply:
Using pflag for now.